src/cbor.rs

```diff
@@ -0,0 +1,136 @@
+            }
+            (Self::UnregisteredIanaValue, Self::UnregisteredIanaValue) => true,
+            (Self::UnregisteredIanaNonPrivateValue, Self::UnregisteredIanaNonPrivateValue) => true,
+            (Self::NonFiniteFloats, Self::NonFiniteFloats) => true,
+            _ => false,
+        }
+    }
+}
+
+impl Eq for CborError {}
+
+impl Clone for CborError {
+    fn clone(&self) -> Self {
+        match self {
+            CborError::DecodeFailed(_) => panic!("cannot clone"),
+            CborError::DuplicateMapKey => CborError::DuplicateMapKey,
+            CborError::EncodeFailed => CborError::EncodeFailed,
+            CborError::ExtraneousData => CborError::ExtraneousData,
+            CborError::OutOfRangeIntegerValue => CborError::OutOfRangeIntegerValue,
+            CborError::UnexpectedItem(msg, want) => CborError::UnexpectedItem(msg, want),
+            CborError::UnregisteredIanaValue => CborError::UnregisteredIanaValue,
+            CborError::UnregisteredIanaNonPrivateValue => {
+                CborError::UnregisteredIanaNonPrivateValue
+            }
+            CborError::NonFiniteFloats => CborError::NonFiniteFloats,
+        }
+    }
+}
+
+impl From<CoseError> for CborError {
+    fn from(e: CoseError) -> Self {
+        match e {
+            CoseError::DecodeFailed(e) => CborError::DecodeFailed(e),
+            CoseError::DuplicateMapKey => CborError::DuplicateMapKey,
+            CoseError::EncodeFailed => CborError::EncodeFailed,
+            CoseError::ExtraneousData => CborError::ExtraneousData,
+            CoseError::OutOfRangeIntegerValue => CborError::OutOfRangeIntegerValue,
+            CoseError::UnexpectedItem(s, s2) => CborError::UnexpectedItem(s, s2),
+            CoseError::UnregisteredIanaValue => CborError::UnregisteredIanaValue,
+            CoseError::UnregisteredIanaNonPrivateValue => {
+                CborError::UnregisteredIanaNonPrivateValue
+            }
+        }
+    }
+}
+
+pub fn to_vec<T>(value: &T) -> Result<Vec<u8>, CborError>
+where
+    T: serde::Serialize,
+{
+    let mut buf = Vec::new();
+    ciborium::into_writer(value, &mut buf)
+        .map_err(coset::CoseError::from)
+        .map_err(CborError::from)?;
+    Ok(buf)
+}
+
+pub fn from_slice<T>(slice: &[u8]) -> Result<T, CborError>
+where
+    T: de::DeserializeOwned,
+{
+    ciborium::from_reader(Cursor::new(&slice))
+        .map_err(|e| CoseError::DecodeFailed(ciborium::de::Error::Semantic(None, e.to_string())))
+        .map_err(CborError::from)
+}
+
+/// Convert a `ciborium::Value` into a type `T`
+#[allow(clippy::needless_pass_by_value)]
+pub fn from_value<T>(value: ciborium::Value) -> Result<T, CborError>
+where
+    T: de::DeserializeOwned,
+{
+    // TODO implement in a way that doesn't require
+    // roundtrip through buffer (i.e. by implementing
+    // `serde::de::Deserializer` for `Value` and then doing
+    // `T::deserialize(value)`).
+    let buf = to_vec(&value)?;
+    from_slice(buf.as_slice())
```

Comment:
You can use [`Value::deserialize`](https://docs.rs/ciborium/0.2.2/ciborium/enum.Value.html#method.deserialized) to avoid round-tripping.